models/helpers.ts

```diff
@@ -130,6 +133,142 @@ export function getLedgerLink(
   };
 }
 
+function getMakeCreditNoteAction(
+  fyo: Fyo,
+  schemaName: ModelNameEnum.SalesInvoice | ModelNameEnum.PurchaseInvoice
+): Action {
+  return {
+    label: fyo.t`Return / Credit Note`,
+    group: fyo.t`Create`,
+    condition: (doc: Doc) =>
+      schemaName === ModelNameEnum.SalesInvoice &&
+      doc.isSubmitted &&
+      !doc.isReturn &&
+      !doc.returnCompleted,
+    action: async (doc: Doc) => {
+      const salesInvoice = doc.getValidDict(true, true) as DocValueMap;
+      const rawCreditNote = (await createReturnDoc(
+        salesInvoice,
+        ModelNameEnum.SalesInvoice,
+        fyo
+      )) as DocValueMap;
+
+      const rawCreditNoteDoc = doc.fyo.doc.getNewDoc(
+        doc.schemaName,
+        rawCreditNote,
+        true
+      ) as Invoice;
+
+      const { openEdit } = await import('src/utils/ui');
+      await openEdit(rawCreditNoteDoc);
+    },
+  };
+}
+export async function updateReturnCompleteStatus(doc: Invoice) {
+  if (!doc.isReturn || !doc.returnAgainst || !doc.items) {
+    return;
+  }
+
+  for (const item of doc.items) {
+    if (!item.item) {
+      return;
+    }
+
+    const invoiceItem = await doc.fyo.db.getAll(`${doc.schemaName}Item`, {
+      filters: {
+        item: item.item,
+        parent: doc.returnAgainst as string,
+      },
+      fields: ['quantity'],
+    });
+
+    if (!invoiceItem) {
+      return;
+    }
+
+    const invoiceItemQty = (invoiceItem[0].quantity as number) ?? 0;
+
+    const returnedItemQty = await doc.fyo.db.getInvoiceReturnedItemQty(
+      doc.schemaName,
+      item.item,
+      doc.returnAgainst as string
+    );
+    if (returnedItemQty !== invoiceItemQty) {
+      return;
+    }
+  }
+  doc.fyo.db.update(doc.schemaName, {
```

Comment:
Don't call `db.update`, you should set the value on the doc and then sync.

Reply:
tried using `setAndSync` .. it didnt work, returned `false` here

https://github.com/frappe/books/blob/b1849928d277f19443b254293e8f4ca2bd5ef9ab/fyo/model/doc.ts#L353-L356

#

should i use `db.update` taking the code below as a reference ?
https://github.com/frappe/books/blob/b1849928d277f19443b254293e8f4ca2bd5ef9ab/models/baseModels/Invoice/Invoice.ts#L140-L144

Reply:
Try setting `null` which is treated as an empty value and `undefined` as a not set value.

If that too isn't working then leave it as it is, it will work as the second case, there will be a sync issue if the doc is in cached. Will be fixed on restart.